pyrelational/pipeline/generic_pipeline.py

```diff
@@ -0,0 +1,312 @@
+        """Returns the performance of the full labelled dataset against the
+        test data. Typically used for evaluation to establish theoretical benchmark
+        of model performance given all available training data is labelled. The
+        "horizontal" line in area under learning curve plots for active learning
+
+        Would not make much sense when we are doing active learning for the real
+        situation, hence not part of __init__
+
+        :param test_loader: Pytorch Data Loader with
+                test data compatible with model, optional as often the test loader can be
+                generated from data_manager but is here for case when it hasn't been defined
+                or there is a new test set.
+
+        :return: performances
+        """
+        self.model.train(self.train_loader, self.valid_loader)
+
+        # use test loader in data_manager if there is one
+        result = self.model.test(self.test_loader if test_loader is None else test_loader)
+        result = self.compute_hit_ratio(result)
+        self.performances["full"] = result
+
+        # make sure that theoretical best model is not stored
+        self.model.current_model = None
+
+        return self.performances["full"]
+
+    def current_performance(self, test_loader: Optional[DataLoader] = None, query: Optional[List[int]] = None) -> Dict:
+        """
+        Current performance of model
+        :param test_loader: Pytorch Data Loader with
+                test data compatible with model, optional as often the test loader can be
+                generated from data_manager but is here for case when it hasn't been defined
+                or there is a new test set.
+
+        :param query: List of indices selected for labelling. Used for calculating hit ratio metric
+        :return: dictionary containing metric results on test set
+
+        """
+        if self.model.current_model is None:  # no AL steps taken so far
+            self.model.train(self.l_loader, self.valid_loader)
+
+        # use test loader in data_manager if there is one
+        result = self.model.test(self.test_loader if test_loader is None else test_loader)
+        result = self.compute_hit_ratio(result, query)
+        return result
+
+    def compute_hit_ratio(self, result: Dict, query: Optional[List[int]] = None) -> Dict:
+        """Utility function for computing the hit ratio as used within the current performance
+        and theoretical performance methods.
+
+        :param result: Dict or Dict-like of metrics
+        :param query: List of indices selected for labelling. Usef for calculating hit ratio metric
+        """
+        if self.data_manager.top_unlabelled is not None:
+            result["hit_ratio"] = (
+                np.nan
+                if query is None
+                else len(set(query) & self.data_manager.top_unlabelled) / len(self.data_manager.top_unlabelled)
```

Comment:
has this been moved to a separate function in a different PR?

Reply:
No this was one of the points you and Alice raised about the 4 lines being repeated in `theoretical_performance` and `current_performance` so it is now its own method called `compute_hit_ratio`

Reply:
ah, right, makes sense! I think one of my comment was about hiding this specific line behind a function name, but this makes more sense !